### PR TITLE
Improve file last modified time detection

### PR DESCRIPTION
**Pull request by avpinchuk**
On some platforms when file copied, `creationTime` attribute of a target file may be updated with current time whereas `lastModifiedTime` attribute of a target file are copying from a source file.

This may leads to issue, when `lastModifiedTime` attribute of a new class file less than `lastModifiedTime` attribute of an old class file, and new class would not be loaded.

Signed-off-by: Alexander Pinčuk <alexander.v.pinchuk@gmail.com>

### CODE REVIEW

**Comment by avpinchuk** [comment]
@arjantijms, please review.

File: src/main/java/org/glassfish/wasp/compiler/Compiler.java [hunk 1]
```diff
@@ -577,6 +580,24 @@ public boolean isOutDated(boolean checkClass) {
...
+     * @return last modified time
+     */
+    private long getLastModifiedTime(File file) {
+        try {
+            BasicFileAttributes basicAttributes = Files.readAttributes(file.toPath(), BasicFileAttributes.class);
+            long creationTime = basicAttributes.creationTime().toMillis();
+            long lastModifiedTime = basicAttributes.lastModifiedTime().toMillis();
+            return Math.max(creationTime, lastModifiedTime);
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (dmatej) [inline src/main/java/org/glassfish/wasp/compiler/Compiler.java:596]: I would prefer using the `log` field.

Comment (avpinchuk) [reply src/main/java/org/glassfish/wasp/compiler/Compiler.java]: Agree. Thank you!

**Review by joakime — COMMENTED** [review]
Some FileSystem and FileTime niggles

File: src/main/java/org/glassfish/wasp/compiler/Compiler.java [hunk 2]
```diff
@@ -577,6 +581,24 @@ public boolean isOutDated(boolean checkClass) {
...
+     * @param file file to read attributes from
+     * @return last modified time
+     */
+    private long getLastModifiedTime(File file) {
+        try {
+            BasicFileAttributes basicAttributes = Files.readAttributes(file.toPath(), BasicFileAttributes.class);
+            long creationTime = basicAttributes.creationTime().toMillis();
+            long lastModifiedTime = basicAttributes.lastModifiedTime().toMillis();
+            return Math.max(creationTime, lastModifiedTime);
+        } catch (IOException e) {
```
[the remark this answers is not in the document]
Comment (joakime) [reply src/main/java/org/glassfish/wasp/compiler/Compiler.java:596]: Also note that you can have FileSystem's that will return one of the 2 values, but not the other (eg: one that returns a lastModifiedTime, but not a creationTime)
That should be taken into account as well.

[the remark this answers is not in the document]
Comment (avpinchuk) [reply src/main/java/org/glassfish/wasp/compiler/Compiler.java:611]: Fixed

File: src/main/java/org/glassfish/wasp/compiler/Compiler.java [hunk 3]
```diff
@@ -577,6 +581,24 @@ public boolean isOutDated(boolean checkClass) {
...
+     *
+     * @param file file to read attributes from
+     * @return last modified time
+     */
+    private long getLastModifiedTime(File file) {
+        try {
+            BasicFileAttributes basicAttributes = Files.readAttributes(file.toPath(), BasicFileAttributes.class);
+            long creationTime = basicAttributes.creationTime().toMillis();
+            long lastModifiedTime = basicAttributes.lastModifiedTime().toMillis();
+            return Math.max(creationTime, lastModifiedTime);
```
[the remark this answers is not in the document]
Comment (avpinchuk) [reply src/main/java/org/glassfish/wasp/compiler/Compiler.java]: Fixed

**Comment by avpinchuk** [comment]
@joakime, thank you for very detailed explanation!

**Review by dmatej — APPROVED** [review]

**Review by arjantijms — APPROVED** [review]

**Pull request merged**